### PR TITLE
Feature: scenario node category verification on validation

### PR DESCRIPTION
**Pull request by lciolecki**

### CODE REVIEW

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/ModelData.scala [hunk 1]
```diff
@@ -94,7 +94,12 @@ trait ModelData extends BaseModelData with AutoCloseable {
   lazy val dictServices: UiDictServices =
     DictServicesFactoryLoader.justOne(modelClassLoader.classLoader).createUiDictServices(processDefinition.expressionConfig.dictionaries, processConfig)
 
-  lazy val validator: ProcessValidator = ProcessValidator.default(processWithObjectsDefinition, dictServices.dictRegistry, modelClassLoader.classLoader)
+  def prepareValidatorForCategory(category: String): ProcessValidator =
```
Comment (mproch) [inline interpreter/src/main/scala/pl/touk/nussknacker/engine/ModelData.scala:97]: maybe let's have category: Option[String]? For most backend tests we don't really care about category?

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/validation/ProcessValidation.scala [hunk 2]
```diff
@@ -9,61 +9,71 @@ import pl.touk.nussknacker.engine.api.expression.ExpressionParser
...
+class ProcessValidation(modelData: ProcessingTypeDataProvider[ModelData],
                         additionalPropertiesConfig: ProcessingTypeDataProvider[Map[String, AdditionalPropertyConfig]],
                         subprocessResolver: SubprocessResolver,
-                        customProcessNodesValidators: ProcessingTypeDataProvider[CustomProcessValidator]) {
+                        customProcessNodesValidators: ProcessingTypeDataProvider[CustomProcessValidator],
+                        expressionParsers: Option[PartialFunction[ExpressionParser, ExpressionParser]]) {
 
   val uiValidationError = "UiValidation"
 
+  private val processValidatorCache = new DefaultCache[String, ProcessValidator](CacheConfig())
```
Comment (mproch) [inline ui/server/src/main/scala/pl/touk/nussknacker/ui/validation/ProcessValidation.scala:41]: this is more tricky than it seems... ProcessingTypeDataProvider can handle model data reload. Can we try to do it as `validators: ProcessingTypeDataProvider[DefaultCache[String, ProcessValidator]]`?

**Review by mproch — COMMENTED** [review]
Looks good, added some minor comments

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala [hunk 3]
```diff
@@ -87,6 +87,7 @@ object DefinitionExtractor {
     // TODO: Use ContextTransformation API to check if custom node is adding some output variable
     def hasNoReturn: Boolean = Set[TypingResult](Typed[Void], Typed[Unit], Typed[BoxedUnit]).contains(returnType)
 
+    def availableForCategory(category: String): Boolean = categories.isEmpty || categories.exists(_.isEmpty) || categories.exists(_.contains(category))
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala:90]: Hmm, some old junk.. ;)

File: ui/server/src/test/scala/pl/touk/nussknacker/ui/integration/DictsFlowTest.scala [hunk 4]
```diff
@@ -55,9 +55,10 @@ class DictsFlowTest extends FunSuite with ScalatestRouteTest with FailFastCirceS
         "label" -> Json.fromString(Label)))
     }
 
-    Get(s"/api/processDefinitionData/${TestProcessingTypes.Streaming}/dict/notExisting/entry?label=fo") ~> addCredentials(credentials) ~> mainRoute ~> checkWithClue {
-      status shouldEqual StatusCodes.NotFound
-    }
+//FIXME: Request was rejected??
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/server/src/test/scala/pl/touk/nussknacker/ui/integration/DictsFlowTest.scala]: My bad - fixed.

**Review by mproch — APPROVED** [review]
:+1:, thanks. I think we should also add entry in MigrationGuide... As we know, there are some users who export/import components from different categories :)

**Pull request merged**